cmd/river/riverbench/river_bench.go

```diff
@@ -0,0 +1,424 @@
+package riverbench
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/signal"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/riverqueue/river"
+	"github.com/riverqueue/river/riverdriver"
+)
+
+type Benchmarker[TTx any] struct {
+	driver       riverdriver.Driver[TTx]
+	duration     time.Duration
+	logger       *slog.Logger
+	name         string
+	numTotalJobs int
+}
+
+func NewBenchmarker[TTx any](driver riverdriver.Driver[TTx], logger *slog.Logger, duration time.Duration, numTotalJobs int) *Benchmarker[TTx] {
+	return &Benchmarker[TTx]{
+		driver:       driver,
+		duration:     duration,
+		logger:       logger,
+		name:         "Benchmarker",
+		numTotalJobs: numTotalJobs,
+	}
+}
+
+// Run starts the benchmarking loop. Stops upon receiving SIGINT/SIGTERM, or
+// when reaching maximum configured run duration.
+func (b *Benchmarker[TTx]) Run(ctx context.Context) error {
+	var (
+		numJobsInserted atomic.Int64
+		numJobsLeft     atomic.Int64
+		numJobsWorked   atomic.Int64
+		shutdown        = make(chan struct{})
+		shutdownClosed  bool
+	)
+
+	// Prevents double-close on shutdown channel.
+	closeShutdown := func() {
+		if !shutdownClosed {
+			b.logger.InfoContext(ctx, "Closing shutdown channel")
+			close(shutdown)
+		}
+		shutdownClosed = true
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Installing signals allows us to try and stop the client cleanly, and also
+	// to produce a final summary log line for th whole bench run (by default,
+	// Go will terminate programs abruptly and not even defers will run).
+	go func() {
+		signalChan := make(chan os.Signal, 1)
+		signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+
+		select {
+		case <-ctx.Done():
+		case <-signalChan:
+			closeShutdown()
+
+			// Wait again since the client may take an absurd amount of time to
+			// shut down. If we receive another signal in the intervening
+			// period, cancel context, thereby forcing a hard shut down.
+			select {
+			case <-ctx.Done():
+			case <-signalChan:
+				fmt.Printf("second signal received; canceling context\n")
+				cancel()
+			}
+		}
+	}()
+
+	if err := b.resetJobsTable(ctx); err != nil {
+		return err
+	}
+
+	workers := river.NewWorkers()
+	river.AddWorker(workers, &BenchmarkWorker{})
+
+	client, err := river.NewClient(b.driver, &river.Config{
+		Logger: slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelWarn})),
+		Queues: map[string]river.QueueConfig{
+			river.QueueDefault: {MaxWorkers: river.QueueNumWorkersMax},
```

Comment:
aaaaWe'd probably want to make this `MaxWorkers` param an argument to the bench command. Setting it to 10k is probably not realistic, and it would probably be useful to play around with different input values to find the sweet spot on throughput.

The fetch cooldown is another attr that's probably worth setting up as an input variable to this. A queue that's optimized for throughput is going to have a much shorter & more aggressive fetch cooldown than most users would typically want (because it would have the tradeoff of constantly hitting the DB to fetch more jobs).

I would suggest setting these each to a specific default that doesn't necessarily map to River's own defaults, because those defaults are likely going to be what's better for a large number of smaller users than the small percentage of users chasing max throughput. But if your purpose is actually to gauge throughput, you'll want some different values.

Reply:
Ah, good call on this one. I chose the max workers after num CPUs performed very poorly, and never reevaluated.

I played around with the numbers quite a bit, and although I wouldn't suggest these are perfect, ended up with some that perform quite well (much better than what I'd had here before). I'm sure more refinement is possible, but will leave that as a future improvement.

``` go
	client, err := river.NewClient(b.driver, &river.Config{
		// When benchmarking to maximize job throughput these numbers have an
		// outside effect on results. The ones chosen here could possibly be
		// optimized further, but based on my tests of throwing a lot of random
		// values against the wall, they perform quite well. Much better than
		// the client's default values at any rate.
		FetchCooldown:     2 * time.Millisecond,
		FetchPollInterval: 5 * time.Millisecond,

		Logger: slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelWarn})),
		Queues: map[string]river.QueueConfig{
			// This could probably use more refinement, but in my quick and
			// dirty tests I found that roughly 1k workers was most optimal. 500
			// and 2,000 performed a little more poorly, and jumping up to the
			// 10k performed considerably less well (scheduler contention?).
			// There may be a more optimal number than 1,000, but it seems close
			// enough to target for now.
			river.QueueDefault: {MaxWorkers: 1_000},
		},
		Workers: workers,
	})
```